### DISCUSSION TITLE
Automatic deep linking

### DISCUSSION THREAD

**Discussion by trin4ik** [expo/router#53] [Ideas]
my opinion, `expo/router` killer-feature is provision "free" deep linking.  for now linking config for `NavigationContainer` generating automatically by [getLinkingConfig](https://github.com/expo/router/blob/c56cd55963988f15da995611af358786fdf167ed/packages/expo-router/src/getLinkingConfig.ts#L68) method. Generates only `path` and `screens` props

### Problem:
no way to set `initialRouteName` to linking config ([react-navigation docs](https://reactnavigation.org/docs/configuring-links/#rendering-an-initial-route)). its needed for some cases:
* when start app with deep link to modal (and you need present modal as modal), solved this in #27, need `initialRouteName` in linking config
* when start app with deep link like `/profile/orders/3` you need `initialRouteName` on each Navigator levels, or you get wrong `back` navigation logic

need to remember, `initialRouteName` in linking config !== `initialRouteName` in Navigator props. they have different logic (hi @nandorojo [react-navigation #10152](https://github.com/react-navigation/react-navigation/issues/10152))

### Solutions (as I can see)
* providing access to modify linking config. i think its not bad practice, but not for this case. i think `expo/router` can and should solve this case automatically
* parsing `initialRouteName` from Navigator and set it to linking options. i tried to do this in #45 pr. the problem is that linking config loads before Navigator rendering. i tried parse component as string, but @EvanBacon [right](https://github.com/expo/router/pull/45#discussion_r986048343), its stupid. but even if we find a way safety prerender or start app only after loads all screens and generated linking config, `initialRouteName` from linking and Navigator still have different logic. i think its bad practice.
* autogenerate `initialRouteName` based on `expo/router` sorted. sounds like solution and all changes we need at [/src/getLinkingConfig.ts](https://github.com/expo/router/blob/3500b4aae8a05e06ec0c5383f8610d5e9aa76a24/packages/expo-router/src/getLinkingConfig.ts#L58):
```diff
- 58     return [node.route, { path, screens }] as const;
+ 58     return [node.route, { path, screens, initialRouteName: node.children[0].route }] as const;
```

its will solve all problems it this case. 
omg, such a simple solution that I don't know why I wrote so much. but in future I think we need tools to merge linking config with custom config.
when i starts wrote this discussion, I thought I'd end with a question: "What do you think about it?". but it seems that the solution came by itself while I was writing 🤷‍♂️

**Comment by AlvSovereign** [comment]
I for one would love to have the linking config exposed.

I'm currently dealing with handling authentication that uses a hash in the deep link URL, which Expo Router does not handle as a search param and therefore cannot find the screen needed to deal with auth tokens.

This was easily achievable via React Navigation which needed a linking prop, where I could customise the config to rewrite the incoming URL to replace the `#` with a `?`.

**Reply by avk-holion** [reply]
Did you find any soloution for this problem?

**Reply by AlvSovereign** [reply]
I am relying on the patch I wrote, here: https://github.com/expo/router/discussions/245#discussioncomment-4887486

This was during the Expo Router RC so I do not know if things have changed in this file, or if there is a new way of editing this config
